src/inset_state/flatten_ellipse_density.cpp

```diff
@@ -0,0 +1,488 @@
+  // time t
+  std::unordered_map<Point, Point> v_intp;
+
+  // (vx_intp_half, vy_intp_half) will be the velocity at the midpoint
+  // (proj_.x + 0.5*delta_t*vx_intp, proj_.y + 0.5*delta_t*vy_intp) at time
+  // t + 0.5*delta_t
+  std::unordered_map<Point, Point> v_intp_half;
+
+  std::unordered_map<Point, Point> velocity;
+
+  // Constants for the numerical integrator
+  const double inc_after_acc = 1.1;
+  const double dec_after_not_acc = 0.75;
+  const double abs_tol = (std::min(lx_, ly_) * 1e-6);
+
+  // Clear previous triangle transformation data
+  proj_qd_.triangle_transformation.clear();
+
+  for (const Point &pt : unique_quadtree_corners_) {
+    proj_qd_.triangle_transformation.insert_or_assign(pt, pt);
+  }
+
+  // TODO: Does the next line adjust rho_mean if the total inset area has
+  //       changed in previous runs of flatten_density() or
+  //       flatten_ellipse_density()?
+  double rho_mean = (1.0 - (initial_area_ / (lx_ * ly_))) /
+                    (1.0 - (total_inset_area() / (lx_ * ly_)));
+
+  // Determine attenuation factor nu that keeps density changes caused by
+  // any ellipse within a fraction f of the mean density.
+  double nu = 0.3;
+
+  std::vector<Ellipse> ells;
+  std::vector<double> ell_density_prefactors;
+  std::vector<double> pwh_rhos;
+  std::vector<double> pwh_areas;
+
+  for (const auto &gd : geo_divs_) {
+    const double rho_p = (target_area_at(gd.id()) / gd.area());
+    for (unsigned int pgon = 0; pgon < gd.n_polygons_with_holes(); ++pgon) {
+      pwh_rhos.push_back(rho_p);
+    }
+  }
+
+  for (auto &gd : geo_divs_) {
+    const double rho_p = (target_area_at(gd.id()) / gd.area());
+    for (unsigned int pgon = 0; pgon < gd.n_polygons_with_holes(); ++pgon) {
+      Polygon_with_holes pwh = gd.polygons_with_holes()[pgon];
+      const double area = pwh_area(pwh);
+      pwh_areas.push_back(area);
+      const Ellipse ell = gd.min_ellipses()[pgon];
+      ells.push_back(ell);
+      ell_density_prefactors.push_back(
+        ellipse_density_prefactor(rho_p, rho_mean, area, nu) /
+        (ell.semimajor * ell.semiminor));
+
+      // Fill GeoDiv name in the same order
+      pgn_id_to_geo_id.push_back(gd.id());
+    }
+  }
+
+  double delta_min = *std::min_element(
+    ell_density_prefactors.begin(),
+    ell_density_prefactors.end());
+  double delta_max = *std::max_element(
+    ell_density_prefactors.begin(),
+    ell_density_prefactors.end());
+
+  std::cerr << "delta_min = " << delta_min << ", delta_max = " << delta_max
+            << std::endl;
+  std::cerr << "rho_mean = " << rho_mean << std::endl;
+  std::cerr << "previous nu = " << nu << std::endl;
+
+  // Update the prefactor densities
+  double acceptable_max = nu * rho_mean;
+  if (delta_max > acceptable_max) {
+    double nu_max = acceptable_max / delta_max;
+    if (nu_max < 1.0) {
+      nu = nu_max;
+    }
+  }
+
+  // find the print the polygons with max delta density
+  int mx_pgn_index = std::max_element(
```

Comment:
Clang-Tidy: Narrowing conversion from 'decltype(__x.base() - __y.base())' (aka 'long') to signed type 'int' is implementation-defined